### ISSUE TITLE
Code execution makes compilation inaccessible by other users

### ISSUE THREAD

**Issue by croessmah** [melpon/wandbox#277]
```
#include <stdlib.h>

int main(int argc, char* argv[])
{
    while(1)
        system(argv[0]);
}
```
All users receive an error "Resource temporarily unavailable" while executing this code.

**Comment by melpon** [comment]
It is difficult to prevent DoS attacks.
Since the execution log is saved, IP blocking by hand will be performed if it interferes with the service operation.

**Comment by croessmah** [comment]
> It is difficult to prevent DoS attacks.
> Since the execution log is saved, IP blocking by hand will be performed if it interferes with the service operation.

DDoS not needed. One run of this code blocks the service for 40-60 seconds.

**Comment by melpon** [comment]
I have no idea resolving this problem.

**Comment by mattgodbolt** [comment]
HI @melpon -- we on the Compiler Explorer team have used some resource limitations to (we think) prevent this issue. Our solution uses `firejail` (we're moving to an equivalent `nsjail` process), and both support cgroups (or cgroup-like) limitations on number of spawned processes (`nsjail` works better though). I'm not 100% sure which of the limits we have in place stops this from being a problem, but we may be able to help. Please let us know if we can be helpful!

(CC @apmorton who has spent a lot of time and energy on our isolation process)

**Comment by melpon** [comment]
Thank you to advice. However, I think if number of processes is limited, the attacker will use concurrent DoS attack if the attacker seriously want to break Wandbox.
Wandbox provides API access, so concurrent access is very easy.

**Comment by mattgodbolt** [comment]
Understood: we would suffer queued compiles in this case too. However we limit the number of concurrent operations so at least serving static web content still works...but compiles would time out and fail too under a concurrent DoS attack. But a single instance of this code doesn't affect us.

Good luck and please don't hesitate to ask if we can help in any way :-)

**Comment by melpon** [comment]
Fixed by #304
